[Protein Browser] Increase PPI graph depth - Part 1: Utils

rogershijin opened the pull request:
This is the first in a series of 3-4 PRs to increase the depth of the protein-protein interaction graph.  I've taken out the PPI graph from the protein browser page temporarily so that these piece-wise changes will not break the page.  All of the important changes are in `data_processing_utils.ts` except one (`GraphDataNested` in `chart.ts`).

- Typing for V1 responses: `V1BaseDatum`, `V1ResponseDatum<T>`, `V1Response<T>`.  @shifucun as the creator of the V1 API, would you mind critiquing these?  (Would greatly appreciate your feedback on the rest of the code too)
- A type `GraphDataNested` which represents a graph as two parallel depthwise nested lists of nodes and links for ease of increasing/decreasing depth (in `chart.ts`).
- General utility functions `objectFromArray` and `zip` inspired by python
- DCID and DCID list manipulation functions `idFromDCID`, `dcidFromID`, `proteinsFromInteractionID`, `sortInteractionDCID`, `deduplicateInteractionDCIDs`, `getInteractionTarget`, `quantityFromDCID`
- Lightweight response data extraction functions `getFromResponse`, `valuesFromResponse`, and `dcidsFromResponse`, which map a response to a list of desired data.
- The heavier response processing function `scoreDataFromResponse` and its helper `symmetricScoreRec`, which together construct a score record mapping interaction IDs to their confidence scores.

Sorry for the extreme slowness on this - I think I find types in typescript more difficult than most people and would greatly appreciate your advice and feedback.  Also if this is still too much code I can split further.

Thanks for the review!

rogershijin commented:
Will revert mixer soon.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,319 @@ export function getProteinDescription(data: GraphNodes): string {
+  const id = interactionID
+    .replace(`${sourceID}_`, "")
+    .replace(`_${sourceID}`, "");
+  if (returnDCID) return dcidFromID(id);
```
For ifs with 1-line bodies, do you guys prefer 1-line or multiple lines?

beets replied:
multiple lines please, for readability.

rogershijin replied:
got it, thanks!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,319 @@ export function getProteinDescription(data: GraphNodes): string {
+export function getInteractionTarget(
+  interactionDCID: string,
+  sourceDCID: string,
+  returnDCID = false
```
Do you guys generally put type annotations on boolean flags? The linter wants this format while codacy wants `returnDCID: boolean = false`.

beets replied:
prefer linter over codacy.

rogershijin replied:
got it, thanks!

rogershijin commented on static/js/protein/chart.ts:
```diff
@@ -58,11 +58,21 @@ export interface InteractionLink extends SimulationLinkDatum<ProteinNode> {
+// nodeDataNested[i] will contain the nodes added in i iterations of BFS,
+// and linkDataNested[i] will contain the links added in i iterations of BFS,
+// along with any links between nodesDataNested[i] and existing nodes in the graph.
+export interface InteractionGraphDataNested {
```
Is `InteractionGraphDataNested` a good name for this? I was also considering `InteractionGraphDataDepthwise`.

rogershijin replied:
Got it that's a great name! And yeah `InteractionGraphData` is the input type to the draw function.

rogershijin replied:
Oh right BFS is just an implementation detail I'll remove it from the description.  And got it I'll collect all the types I use in a single file then.  

Unfortunately I'm doing BFS from the client right now.  I'll PR this soon for you guys to look at so we can decide what should be moved to flask.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -12,7 +13,39 @@ import { ProteinVarType } from "./page";
 
-const MAX_INTERACTIONS = 10; // upper bound on node degree in interaction graph viz's
+// Base type of "values" value of objects stored in <V1 response>.data.data
+type BaseDCDataType = {
```
@shifucun all of these properties guaranteed in every response?

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,319 @@ export function getProteinDescription(data: GraphNodes): string {
+  const scoreList = scoreValues
+    // filter results for IntactMiScore - there should be 0 or 1 match
+    .map((scoreRecList: BaseDCDataType[] | undefined) => {
+      if (scoreRecList !== undefined) {
```
Codacy doesn't like this check for if `scoreRecList` is `undefined`.  What's the proper way to do things?  Thanks!

chejennifer replied:
you could just check `!_.isEmpty(scoreRecList)`

rogershijin replied:
Thanks! I'll do this

rogershijin replied:
I'm sure you've experienced more weird null cases than I have in your lifetime, so I'm gonna add it!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,324 @@ export function getProteinDescription(data: GraphNodes): string {
+/**
+ * Given id of the form {protein id}_{species id} (e.g. P53_HUMAN), parse into and return ProteinNode
+ */
+export function nodeFromID(
```
Previously checked in (moving to resolve codacy error)

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,324 @@ export function getProteinDescription(data: GraphNodes): string {
+
+      }.
+ */
+export function getProteinInteractionGraphData(
```
Previously checked in (moving to resolve codacy error)

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,324 @@ export function getProteinDescription(data: GraphNodes): string {
+    };
+  });
+
+  return {
```
Only big change is return type is now `InteractionGraphDataNested` instead of `InteractionGraphData`

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -12,7 +13,32 @@ import { ProteinVarType } from "./page";
+  name: string;
+  provenanceId: string;
+  types: string[];
+};
```
@shifucun Should some of these properties be optional? Thanks!

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
+ */
+// Recursive elementwise type inference on array of arrays
+type Zip<T> = T extends [Array<infer U>, ...infer Z] ? [U, ...Zip<Z>] : [];
+export function zip<T extends unknown[][]>(...arrays: T): Array<Zip<T>> {
```
i don't think you need to create your own zip method since lodash has one: https://lodash.com/docs/4.17.15#zip

rogershijin replied:
Oh oops thanks didn't know! Will just delete this and use lodash's then unless I find an example for which this does better type inference than lodash.

rogershijin commented on static/js/protein/chart.ts:
```diff
@@ -783,103 +793,103 @@ export function drawProteinInteractionChart(
-    .text(({ name }) => name)
-    .attr("class", "protein-node-label");
-}
+// export function drawProteinInteractionGraph(
```
Oh that's right I think I was getting type errors I was too lazy to deal with when I first commented this out but perhaps now that the code is self-consistent again they'll go away.

rogershijin replied:
Yup you're right I was able to uncomment with a minor tweak to the input type.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -12,7 +13,32 @@ import { ProteinVarType } from "./page";
+};
+
+// Generic for objects stored in <V1 response>.data.data
+type V1ResponseDatum<ValueDatum extends V1BaseDatum> = {
```
Oh yeah I think I need to ask @shifucun about this.  It might be that everything returns stuff of type `V1BaseDatum` and we don't need a generic.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,324 @@ export function getProteinDescription(data: GraphNodes): string {
+  // danger: assumes neither {protein id}, {species id} contain an underscore
+  const split = id.split("_");
+  if (split.length !== 4) {
+    throw `Unsupported DCID ${interactionDCID}`;
```
Oh yeah failing loudly might not be the best I'll just do a `console.assert`.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,324 @@ export function getProteinDescription(data: GraphNodes): string {
+  // so we need to keep track of which order is the true DCID corresponding to an entity in the response.
+  // Counterexample: http://datacommons.org/browser/bio/KLOTB_HUMAN
+  return Object.values(
+    objectFromArray(interactionDCIDs, sortInteractionDCID, (dcid) => dcid)
```
I think python has brainwashed me into not finding your way simpler but I trust your judgment/will implement this insteaad.

rogershijin replied:
Sorry about the helpers guys I think I have bad intuitions about what other people find readable.  This is very useful feedback.  

I'll reimplement with a loop and not try to be fancy.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,324 @@ export function getProteinDescription(data: GraphNodes): string {
+  scores: number[]
+): Record<string, number> {
+  const scoreRec = {};
+  zip(interactionDCIDs, scores).forEach(([dcid, score]) => {
```
Sorry yes totally unnecessary! I think this is a bad combo of when I code in js I sometimes forget for loops exist/my python habits making me allergic to array indexing.

Will fix thanks for pointing this out!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,328 @@ export function getProteinDescription(data: GraphNodes): string {
+ * Given an object mapping interaction IDs to confidence scores and the DCIDs of two proteins, return the score
+ * of the interaction between the two proteins if it appears in the object, or the default score otherwise.
+ */
+export function scoreFromProteinDCIDs(
```
Hmm I'm still ironing out the refactor of the downstream things but I currently believe it'll only be used once so it can be taken out.

rogershijin replied:
Deleted!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +537,328 @@ export function getProteinDescription(data: GraphNodes): string {
+  return resp.data.data.map((obj) => obj[key]);
+}
+
+// Useful special cases of getFromResponse
```
Thanks noted! I'll replace with the calls to the general function.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
+  const interactionDCIDs = dcidsFromResponse(scoreResponse);
+  const scoreList = scoreValues
+    // filter results for IntactMiScore - there should be 0 or 1 match
+    .map((scoreRecList: V1BaseDatum[] | undefined) => {
```
Whoa keen eye thanks for catching! Yes I also think it can be done in one.

beets commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
+ * Convert DCID of form bio/<id> to id.
+ * Note: probably should be doing dcid: `bio/${string}`, but currently causes many type errors
+ */
+export function idFromDCID(dcid: string): string {
```
Since there are so many ID types floating around in this module -- perhaps specify that this is the `proteinId`? I see places here where this sometimes include species -- specificity will go a long way towards reducing errors.

Also - only export functions that will be used outside the module (applies to everything else in here).

rogershijin replied:
Oh ok thanks for pointing this out - the only two ID types for the data of the interaction graph which are true IDs (i.e. used to ID nodes/links in the graph) are protein ID and interaction ID, and `idFromDCID` and `dcidFromID` apply to both.  However in `chart.ts` there's stuff like `elementID`, `barIDFunc`, `circleIDFunc`.  Perhaps I could make `idFromDCID` and `dcidFromID` static methods in the new interaction graph component?

Also I think I wrote species ID in a comment somewhere without realizing it was confusing - I'll just revise the comments so there's no mention of IDs in the graph part other than protein ID and interaction ID.

rogershijin replied:
I renamed these functions to `ppiIDFromDCID` and `ppiDCIDFromID` and replaced all mentions of `species ID` with `species name`.  

To make this a bit safer I also added an assertion in `ppiIDFromDCID` to enforce that the input cannot start with `"bio/"`, and a new type `bioDCID = `bio/${string}` to enforce that all variables labeled DCID throughout the code start with `"bio/"`.  (please let me know if you think this is overkill I can revert back to just string type for everything).

beets commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
+/**
+ * Given interaction DCID of the form bio/{protein1}_{protein2}, return [protein1, protein2]
+ */
+export function proteinsFromInteractionDCID(interactionDCID: string): string[] {
```
Do we get access to the underlying nodes for these? That can avoid the assumptions for the structure of these dcids.

Would be helpful to have an example in the comments here - e.g. bio/2A5D_YEAST_AHA1_YEAST

rogershijin replied:
I've asked Bo a similar question before and I think this would require another hop in the knowledge graph (which from the clientside would be another request).  Migrating to flask and using the [interactionProtein](https://datacommons.org/browser/bio/1433G_HUMAN_P53_HUMAN) property would be much cleaner than what I'm currently doing.  

This might be overkill but another request might not be too bad because it could be sent in parallel with the request for interaction scores. `interactinType` and `interactionDetectionMethod` might also be useful to pull. (but yes this would all be better in flask)

beets commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
+ * Given interactionDCID of the form {protein 1}_{protein 2}, return ID of the same form
+ * but with protein 1, protein 2 in alpha order.
+ */
+function sortInteractionDCID(interactionDCID: string): string {
```
Since this is only used for the following function - this could return a `proteins` to avoid the string munging.

rogershijin replied:
Oh do you mean something like `[P53_HUMAN, 1433G_HUMAN]`?  If so I think we can delete this function and add a `sorted` flag to `proteinsFromInteractionDCID`.

rogershijin replied:
I ended up deleting this helper without adding the `sorted` flag because all use cases so far would work with `sorted=False`.

beets reviewed (commented):
I feel the code will be easier if this was encapsulated in a component.. then you can store some of the intermediate data and avoid so many string operations.

please add tests for the helpers (or please be sure to do so in follow ups, since the code might change after the series of PRs).

thanks for the detailed comments in the code. i also recognize the amount of time you must have spent figuring out all the typing issues and typescript <> python translations. this is part of the reason smaller, earlier pr's are helpful so we can align on coding practices.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -170,94 +196,6 @@ export function getProteinInteraction(
-/**
- * Given id of the form {protein id}_{species id} (e.g. P53_HUMAN), parse into and return ProteinNode
- */
-function nodeFromID(proteinSpeciesID: string, depth: number): ProteinNode {
```
yeah these were just to move this function and the one after - sorry didn't know it would result in a confusing PR view

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
+/**
+ * Given quantity DCID of the form "<quantityName><quantityValue>", extract and return quantityValue.
+ */
+export function quantityFromDCID(
```
great name will adopt!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -1,8 +1,9 @@
+import { AxiosResponse } from "axios";
```
Thank you so much for the review! Yes will do.

beets reviewed (commented):
thanks for the updates -- there's a huge improvement in readability. a few more comments / questions below.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -1,18 +1,44 @@
+/**
+ * Copyright 2021 Google LLC
```
Thanks for reviewing again!! And for eagle-eyeing this

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,263 @@ export function getProteinDescription(data: GraphNodes): string {
+/**
+ * Convert DCID of form bio/<id> to id. Utility for protein-protein interaction (PPI) graph.
+ */
+export function ppiIDFromDCID(dcid: bioDCID): string {
```
got it! will update everywhere!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,263 @@ export function getProteinDescription(data: GraphNodes): string {
+/**
+ * Given id, convert to DCID of form bio/<id>.  Utility for protein-protein interaction (PPI) graph.
+ */
+export function ppiDCIDFromID(id: string): bioDCID {
```
expunged!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,263 @@ export function getProteinDescription(data: GraphNodes): string {
+  quantityName: string
+): number {
+  if (!quantityDCID.includes(quantityName)) {
+    throw `DCID "${quantityDCID}" does not contain quantity name "${quantityName}"`;
```
got it! returned `NaN`!

beets replied:
don't forget to check for NaN downstream (and please document this as a possible value)

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,263 @@ export function getProteinDescription(data: GraphNodes): string {
+  const centerNodeID = data[0].parent;
+  let neighbors = data.map(({ name: interactionID, value }) => {
+    // value is confidenceScore
+    console.log("id", interactionID);
```
sorry about this I've become sloppy

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,263 @@ export function getProteinDescription(data: GraphNodes): string {
+/**
+ * Given an array of interaction DCIDs and a parallel array of corresponding scores,
+ * construct and return score record such that for each interaction A_B with corresponding DCID in interaction DCIDs,
+ * both A_B and B_A map to the score of A_B.
```
Oh good question!  I'm not sure this is the right thing to do, but we currently do not graphically distinguish between the A_B and B_A scores, and Sam has said that A_B.score should equal B_A.score.  But in the future perhaps we should make the graph directed to make relationships like "A phosphorylates B" more clear.  I also think there should be assays for which the scores are not symmetric - ex. "knockout A and see what happens to the expression of B".  These might not exist in the graph currently but could very well be imported in the future.

I think the treatment of A_B vs. A_B will require more careful thinking later on, but for now I'll add something which keeps the higher of the two scores.

rogershijin replied:
Sure will add!

rogershijin commented on static/js/protein/types.ts:
```diff
@@ -0,0 +1,79 @@
+import { AxiosResponse } from "axios";
+import { SimulationLinkDatum, SimulationNodeDatum } from "d3";
+
+export type bioDCID = `bio/${string}`;
```
Since this applies to all dcids across biomedical data commons, how about "biomedDcid"? (or if we want something specific to the interaction graph, we could do "bioNodeOrInteractionDcid")

rogershijin replied:
Cool thanks!

rogershijin commented on static/js/protein/types.ts:
```diff
@@ -0,0 +1,79 @@
+
+// interfaces for protein-protein interaction graph
+
+export interface Node {
```
Oh this was intended to be a broad interface (`ProteinNode` below extends it).  Some other proposed projects on datacommons (drug browser/Crystal's project) might use parts of this code for graph viz so wanted to eventually factor out the graph stuff to an interface totally independent of protein stuff, and make proteins a special case.  What do you think?

rogershijin commented on static/js/protein/types.ts:
```diff
@@ -0,0 +1,80 @@
+  name: string;
+  provenanceId: string;
+  types: string[];
+}
```
I was really tempted to do 

```
export interface V1BioDatum extends Omit<Required<V1BaseDatum>, "value"> {
  dcid: bioDcid;
}
```

here to avoid copying

```
  name: string;
  provenanceId: string;
  types: string[];
```

but held back for the sake of readability.  Hope the reviewers are proud :p

rogershijin commented:
ptal! Changes:
- IDs/DCIDs are now properly camelCased to Ids/Dcids
- `symmetricScoreRec` now assigns `max(A_B.score, B_A.score)` to both `A_B` and `B_A`
- type `V1BaseDatum` now has an optional `value` property (thanks @shifucun for comments and [this example](https://autopush.api.datacommons.org/v1/property/out/name/values/country/USA)!)
- `quantityFromDcid` now returns `NaN` when it fails to parse.

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -12,7 +13,32 @@ import { ProteinVarType } from "./page";
+
+// Generic for V1 response
+// Reference: https://github.com/axios/axios/blob/7d6bddba2d8de29c263feaef4c40daa50cb4b176/index.d.ts#L83
+type V1Response<ValueDatum extends V1BaseDatum> = AxiosResponse<{
```
Yes, but you can just make the type 
```
interface V1Response {
data: ...
}
```
and functions that take in the V1 response would have an argument of type `V1Response` and where these functions are called, just pass `resp.data`?

rogershijin replied:
Oh ok sure this works and I like that it is independent of Axios.  I'll make the change!

chejennifer reviewed (commented):
thanks for the updates

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,254 @@ export function getProteinDescription(data: GraphNodes): string {
+  interactionDcid: bioDcid,
+  sourceDcid: bioDcid,
+  returnDcid = false
+): typeof returnDcid extends true ? bioDcid : string {
```
yeah that would be cleaner, thanks!  I also realized I never verified that this return type actually works.

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,254 @@ export function getProteinDescription(data: GraphNodes): string {
+ * Given id, convert to DCID of form bio/<id>.  Utility for protein-protein interaction (PPI) graph.
+ */
+export function ppiDcidFromId(id: string): bioDcid {
+  console.assert(!id.includes("bio/"), `Invalid ID: ${id}`);
```
I'll check with Sam and get back to you, thanks for the detailed instructions covering what to do in all the cases!

rogershijin replied:
Ok Sam said it is important to communicate these errors to the user but ideally in the form of a log file they can download.  But for now we'll log these to the console and the user can download the console logs.

rogershijin replied:
But I think this location is not an informative place to do it so I'll delete the assert here.

rogershijin replied:
Ok I think I'm also not doing good enough error handling to be able to tell the users I'm doing something, so I'll just take out all of these asserts

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,254 @@ export function getProteinDescription(data: GraphNodes): string {
+    const proteins = proteinsFromInteractionDcid(interactionDcid);
+    if (!interactions.has(proteins)) {
+      uniqueInteractionDcids.push(interactionDcid);
+
```
expunged!

rogershijin replied:
oh ahh sorry! I think I did this in the wrong branch.  definitely gone now!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,254 @@ export function getProteinDescription(data: GraphNodes): string {
+  scores: number[]
+): Record<string, number> {
+  const scoreRec = {};
+  console.assert(interactionDcids.length === scores.length);
```
ok! will take this out

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,254 @@ export function getProteinDescription(data: GraphNodes): string {
+    const [protein1, protein2] = proteinsFromInteractionDcid(
+      interactionDcids[i]
+    );
+    const score12 = _.get(
```
yea makes sense!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,254 @@ export function getProteinDescription(data: GraphNodes): string {
+  const interactionDcids = getFromResponse(scoreResponse, "entity");
+  const scoreList = scoreValues
+    // filter results for IntactMiScore - there should be 0 or 1 match
+    .map((bioData: V1BioDatum[] | undefined) => {
```
got it! will take out

rogershijin commented on static/js/protein/types.ts:
```diff
@@ -0,0 +1,79 @@
+}
+
+// Generic for objects stored in <V1 response>.data.data
+export type V1ResponseDatum<ValueDatum extends V1BaseDatum> = {
```
thanks for thinking carefully about this! I'll check with Bo.

about the naming thing though - I think I'm calling mixer directly so these types were intended just for the mixer response.  is there a requirement for flask endpoints to follow the same or a similar format?

chejennifer replied:
I don't think you should be calling mixer directly..  and what format are you referring to?

rogershijin replied:
Oh ok perhaps I'm confused.  The root endpoint is https://autopush.api.datacommons.org/v1/bulk . Is this flask?

Also sorry I was referring to the flask response format.  Supposing that I'm calling mixer directly and the types are then for mixer responses, would they also work for flask responses?

rogershijin replied:
Oh Bo agrees with your advice! I also think that extending an interface with only optional properties is no better than extending `{}`.  Each user would still have to make their own subinterfaces which would repeat subsets of the keys.

rogershijin replied:
Ah ok got it thanks.  I do hope to migrate a lot of the chained data retrieval/processing to flask eventually.  I guess I'll keep the V1s in the names and then eventually redo/rename these types when I migrate to flask?

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +530,254 @@ export function getProteinDescription(data: GraphNodes): string {
+      }
+      // if 1 match, return the retrieved IntactMiScore
+      console.assert(emptyOrInteractionSingleton.length === 1);
+      return quantityFromDcid(
```
this turned out much much cleaner than what I had! thanks for the simplification!

rogershijin commented:
[ptal!] Thanks @chejennifer and @beets for another review! Changes:
- simplified response types following @chejennifer's design - in particular `V1Response` is no longer generic (and has been renamed to `V1BioResponse`).  
- removed `console.assert`s in all but one place where I check for NaN values in `scoreDataFromResponse`
- Updated comments to reflect your suggestions

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,248 @@ export function getProteinDescription(data: GraphNodes): string {
+  scores: number[]
+): Record<string, number> {
+  const scoreRec = {};
+  for (let i = 0; i < interactionDcids.length; i++) {
```
in case the two arrays are different lengths, you want the termination condition to be `i < Math.min(interactionDcids.length, scores.length)`

rogershijin replied:
ah ok that's right, thanks for catching!

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,248 @@ export function getProteinDescription(data: GraphNodes): string {
+  const scoreValues = getFromResponse(scoreResponse, "values");
+  const interactionDcids = getFromResponse(scoreResponse, "entity");
+  const scoreList = scoreValues
+    // map scoreValues to IntactMiScore if available or default interaction score
```
".. if available, otherwise map to default interaction score"

rogershijin replied:
thanks for catching! done!

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,248 @@ export function getProteinDescription(data: GraphNodes): string {
+          if (!isNaN(score)) {
+            return score;
+          }
+          console.warn(`Invalid quantity ID ${bioDatum.dcid} -- skipping`);
```
I'm not sure putting the console.warn here makes the most sense either because you're still assigning a score here (DEFAULT_INTERACTION_SCORE), so for a user, it won't actually look like anything has been skipped? 

Maybe the console.warn can be added where this function is called and we can console.warn all cases where the score is DEFAULT_INTERACTION_SCORE & the message can be something like `"Unable to retrieve score for interaction <interaction dcid> -- default interaction score of <interaction score value> used"`

rogershijin replied:
Ohhh! Yes that's much better. I changed it to

```
scoreValues
    .map((bioData: V1BioDatum[], i) => {
      for (const bioDatum of bioData) {
        ...
      }
      console.warn(`Unable to retrieve score for interaction ${interactionDcids[i]} -- default score of ${DEFAULT_INTERACTION_SCORE} used`);
      return DEFAULT_INTERACTION_SCORE;
    })
```

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,248 @@ export function getProteinDescription(data: GraphNodes): string {
+  const id = ppiIdFromDcid(interactionDcid);
+  // danger: assumes neither {protein name}, {species name} contain an underscore
+  const split = id.split("_");
+  return [`${split[0]}_${split[1]}`, `${split[2]}_${split[3]}`];
```
got it! will do!

rogershijin commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,248 @@ export function getProteinDescription(data: GraphNodes): string {
+  resp: V1BioResponse,
+  key: K
+): V1BioResponseDatum[K][] {
+  if (!("data" in resp.data)) {
```
Oh that's right! great catch! done!

rogershijin commented:
[ptal!] Thanks for all the reviews so far! Changes:
- Improve user warnings in `scoreDataFromResponse` according to @chejennifer's comments
- `proteinsFromInteractionDcid` now returns null when interaction dcid does not contain exactly 3 underscores
- `symmetricScoreRec` and `deduplicateInteractionDcids`, which call `proteinsFromInteractionDcid`, now check for nullity and skip offending interaction dcids

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,272 @@ export function getProteinDescription(data: GraphNodes): string {
+): Record<string, number> {
+  const scoreValues = getFromResponse(scoreResponse, "values");
+  const interactionDcids = getFromResponse(scoreResponse, "entity");
+  if (scoreValues.length !== interactionDcids.length) {
```
Since we care about the scoreValues and interactionDcids together, I think it will be cleaner to just loop through scoreResponse instead of extracting out scoreValues and interactionDcids. If interactionDcids are unique, I would also suggest passing a record with interactionDcid as key and score as value to the symmetricScoreRec function.

So we would have something like: 
```
const scores = {}
for (const responseDatum of scoreResponse.data) {
  const interactionDcid = responseDatum.entity;
  const values = responseDatum.values;
  for (const value of values) {
     ... try to get score, if there is a score add it to the record and break
   }
  warn about no score and add to the record
}
```

rogershijin replied:
oh yes this is great I agree this would be a lot cleaner, thanks for the pseudocode!

chejennifer reviewed (commented):
thanks for the updates, one more suggestion

rogershijin commented:
[ptal!] Thanks @chejennifer for looking at this so many times! Made the changes in your latest comment!

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,271 @@ export function getProteinDescription(data: GraphNodes): string {
+  for (const responseDatum of scoreResponse.data) {
+    const interactionDcid = responseDatum.entity;
+    if (!interactionDcid) {
+      console.warn("Undefined or empty interaction DCID -- skipping");
```
include what the interaction DCID that's being skipped is?

rogershijin replied:
done!

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,271 @@ export function getProteinDescription(data: GraphNodes): string {
+    }
+    const interactionId = ppiIdFromDcid(interactionDcid);
+    const values = responseDatum.values;
+    for (const bioDatum of values || []) {
```
move `|| []` to when defining `values`

rogershijin replied:
done!

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,271 @@ export function getProteinDescription(data: GraphNodes): string {
+        );
+        if (!isNaN(score)) {
+          scoreRec[interactionId] = score;
+          continue;
```
we probably want to `break` when a score is found

rogershijin replied:
oops yes absolutely right!

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,271 @@ export function getProteinDescription(data: GraphNodes): string {
+          continue;
+        }
+      }
+      console.warn(
```
I think this needs to go outside of the current for loop, we don't want to do this for every bioDatum that isn't a score. Should also first check that no score was found for the interactionId

rogershijin replied:
oh yes my mistake! thank you so much for catching this

chejennifer commented on static/js/protein/data_processing_utils.ts:
```diff
@@ -592,3 +529,271 @@ export function getProteinDescription(data: GraphNodes): string {
+    scoreRec[`${proteinA}_${proteinB}`] = maxScore;
+    scoreRec[`${proteinB}_${proteinA}`] = maxScore;
+  }
+  return scoreRec;
```
usually it's better to avoid mutating the object that gets passed in, so here I would create a new record object that holds the new values

rogershijin replied:
oh yea I also like pure functions. done!

beets reviewed (approved):
thanks for all the iterations! lgtm, pending jennifer's approval

chejennifer reviewed (approved):
thanks for all the changes!

rogershijin commented:
Thank you so much for the reviews/continued guidance! The code is much cleaner now than before.
